flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkUtils.java

```diff
@@ -108,21 +116,109 @@ private static void mergeInto(JsonNode toNode, JsonNode fromNode) {
+            boolean deleteHaConfigmaps) {
         kubernetesClient
                 .apps()
                 .deployments()
                 .inNamespace(namespace)
-                .withName(clusterId)
+                .withName(KubernetesUtils.getDeploymentName(clusterId))
                 .cascading(true)
                 .delete();
+
+        if (deleteHaConfigmaps) {
+            // We need to wait for cluster shutdown otherwise confimaps might be recreated
+            waitForClusterShutdown(kubernetesClient, namespace, clusterId);
+            kubernetesClient
+                    .configMaps()
+                    .inNamespace(namespace)
+                    .withLabels(
+                            KubernetesUtils.getConfigMapLabels(
+                                    clusterId, LABEL_CONFIGMAP_TYPE_HIGH_AVAILABILITY))
+                    .delete();
+        }
+    }
+
+    /** We need this due to the buggy flink kube cluster client behaviour for now. */
+    public static void waitForClusterShutdown(
+            KubernetesClient kubernetesClient, String namespace, String clusterId) {
+
+        boolean jobManagerRunning = true;
+        boolean serviceRunning = true;
+
+        for (int i = 0; i < 60; i++) {
+            if (jobManagerRunning) {
+                PodList jmPodList =
+                        kubernetesClient
+                                .pods()
+                                .inNamespace(namespace)
+                                .withLabel(
+                                        Constants.LABEL_TYPE_KEY, Constants.LABEL_TYPE_NATIVE_TYPE)
+                                .withLabel(
+                                        Constants.LABEL_COMPONENT_KEY,
+                                        Constants.LABEL_COMPONENT_JOB_MANAGER)
+                                .withLabel(Constants.LABEL_APP_KEY, clusterId)
+                                .list();
+
+                if (jmPodList.getItems().isEmpty()) {
+                    jobManagerRunning = false;
+                }
+            }
+
+            if (serviceRunning) {
+                Service service =
+                        kubernetesClient
+                                .services()
+                                .inNamespace(namespace)
+                                .withName(
+                                        ExternalServiceDecorator.getExternalServiceName(clusterId))
+                                .fromServer()
+                                .get();
+                if (service == null) {
+                    serviceRunning = false;
+                }
+            }
+
+            if (!jobManagerRunning && !serviceRunning) {
+                break;
+            }
+            LOG.info("Waiting for cluster shutdown... ({})", i);
+            try {
+                Thread.sleep(1000);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
```

Comment:
It would be good to log that shutdown was completed, otherwise the last thing we see is:
```2022-02-28 21:12:45,381 o.a.f.k.o.u.FlinkUtils         [INFO ] [default.basic-example] Waiting for cluster shutdown... (5)
2022-02-28 21:12:46,395 o.a.f.k.o.u.FlinkUtils         [INFO ] [default.basic-example] Waiting for cluster shutdown... (6)
```